plugins/experimental/certifier/certifier.cc

```diff
@@ -0,0 +1,719 @@
+  TSMutex list_mutex;
+
+  int size = 0;
+  int limit;
+  SslData *head = nullptr;
+  SslData *tail = nullptr;
+
+public:
+  SslLRUList(int in_limit = 4096) : limit(in_limit) { list_mutex = TSMutexCreate(); }
+
+  ~SslLRUList() { TSMutexDestroy(list_mutex); }
+
+  // Returns valid ptr to SSL_CTX if successful lookup
+  //         nullptr if not found and create SslData in the map
+  SSL_CTX *
+  lookup_and_create(const char *servername, void *edata, bool &wontdo)
+  {
+    SslData *ssl_data              = nullptr;
+    scoped_SslData scoped_ssl_data = nullptr;
+    SSL_CTX *ref_ctx               = nullptr;
+    std::string commonName(servername);
+    TSMutexLock(list_mutex);
+    auto dataItr = cnDataMap.find(commonName);
+    /// If such a context exists in dict
+    if (dataItr != cnDataMap.end()) {
+      /// Reuse context if already built, self queued if not
+      if ((ssl_data = dataItr->second.get())->wontdo) {
+        wontdo = true;
+      } else if (ssl_data->ctx) {
+        ref_ctx = ssl_data->ctx.get();
+      } else {
+        ssl_data->vconnQ.push(edata);
+      }
+    } else {
+      /// Add a new ssl_data to dict if not exist
+      scoped_ssl_data.reset(new SslData);
+      ssl_data             = scoped_ssl_data.get();
+      ssl_data->commonName = std::move(commonName);
+      ssl_data->vconnQ.push(edata);
+
+      auto &target_data = cnDataMap[ssl_data->commonName];
+
+      if (target_data == nullptr) {
+        target_data = std::move(scoped_ssl_data);
+      } else {
+        TSDebug(PLUGIN_NAME, "something wrong, stringview?");
+      }
+    }
+    // With a valid sslData pointer
+    if (ssl_data != nullptr) {
+      // Add to the list and set scheduled flag
+      prepend(ssl_data);
+      if (ref_ctx == nullptr || !ssl_data->scheduled) {
+        ssl_data->scheduled = true;
+      }
+    }
+    TSMutexUnlock(list_mutex);
+    return ref_ctx;
+  }
+
+  // Setup ssldata 1) ctx 2) cert 3) swapping queue
+  // Ownership of unique pointers are transferred into this function
+  // Then if the entry is found, the ownership is further transferred to the entry
+  // if not, the objects are destroyed here. (As per design, this is caused by LRU management deleting oldest entry)
+  void
+  setup_data_ctx(const std::string &commonName, std::queue<void *> &localQ, std::unique_ptr<SSL_CTX> ctx,
+                 std::unique_ptr<X509> cert, const bool &wontdo)
+  {
+    TSMutexLock(list_mutex);
+    auto iter = cnDataMap.find(commonName);
+    if (iter != cnDataMap.end()) {
+      std::swap(localQ, iter->second->vconnQ);
+      iter->second->ctx    = std::move(ctx);
+      iter->second->cert   = std::move(cert); ///< We might not need cert, can be easily removed
+      iter->second->wontdo = wontdo;
+    }
+    TSMutexUnlock(list_mutex);
+  }
+
+  // Prepend to the LRU list
+  void
+  prepend(SslData *data)
+  {
+    TSMutexLock(list_mutex);
+    std::unique_ptr<SslData> local = nullptr;
+    if (data != nullptr) {
```

Comment:
Why not call `remove` to remove the item, then always insert at the front?

Reply:
the remove method does removal of both list node and dict node. Here in `prepend` the first part is only removing from list.

Reply:
It might be possible to write a method/function that just did the part of remove that's common to both.